File: .taira-source-c80bf4d85a97/java/iroha_android/src/main/java/org/hyperledger/iroha/android/client/RamLfeOutputOpening.java
package org.hyperledger.iroha.android.client;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/** Externally attested opening of a RAM-LFE encrypted output. */
public final class RamLfeOutputOpening {
  private final RamLfeOutputOpeningPayload payload;
  private final String signature;

  public RamLfeOutputOpening(final RamLfeOutputOpeningPayload payload, final String signature) {
    this.payload = Objects.requireNonNull(payload, "payload");
    this.signature = Objects.requireNonNull(signature, "signature");
  }

  public RamLfeOutputOpeningPayload payload() {
    return payload;
  }

  public String signature() {
    return signature;
  }

  Map<String, Object> toJsonMap() {
    final Map<String, Object> opening = new LinkedHashMap<>();
    opening.put("payload", payload.toJsonMap());
    opening.put("signature", HttpClientTransport.normalizeEvenLengthHex(signature, "opening.signature"));
    return opening;
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/client/RamLfeOutputOpening.kt
package org.hyperledger.iroha.sdk.client

import java.util.LinkedHashMap

/** Canonical payload signed by an external RAM-LFE output-opening authority. */
class RamLfeOutputOpeningPayload(
    @JvmField val programId: String,
    @JvmField val inputCiphertextHash: String,
    @JvmField val outputCiphertextHash: String,
    @JvmField val parameterDigest: String,
    @JvmField val evaluationKeyDigest: String,
    @JvmField val openedOutputHash: String,
    @JvmField val openedAtMs: Long,
    @JvmField val expiresAtMs: Long?,
) {
    fun toJsonMap(): Map<String, Any> {
        val payload = LinkedHashMap<String, Any>()
        payload["program_id"] = HttpClientTransport.normalizeNonBlank(programId, "opening.payload.programId")
        payload["input_ciphertext_hash"] =
            HttpClientTransport.normalizeHex32(inputCiphertextHash, "opening.payload.inputCiphertextHash")
        payload["output_ciphertext_hash"] =
            HttpClientTransport.normalizeHex32(outputCiphertextHash, "opening.payload.outputCiphertextHash")
        payload["parameter_digest"] =
            HttpClientTransport.normalizeHex32(parameterDigest, "opening.payload.parameterDigest")
        payload["evaluation_key_digest"] =
            HttpClientTransport.normalizeHex32(evaluationKeyDigest, "opening.payload.evaluationKeyDigest")
        payload["opened_output_hash"] =
            HttpClientTransport.normalizeHex32(openedOutputHash, "opening.payload.openedOutputHash")
        payload["opened_at_ms"] = openedAtMs
        if (expiresAtMs != null) {
            payload["expires_at_ms"] = expiresAtMs
        }
        return payload
    }
}

/** Externally attested opening of a RAM-LFE encrypted output. */
class RamLfeOutputOpening(
    @JvmField val payload: RamLfeOutputOpeningPayload,
    @JvmField val signature: String,
) {
    fun toJsonMap(): Map<String, Any> {
        val opening = LinkedHashMap<String, Any>()
        opening["payload"] = payload.toJsonMap()
        opening["signature"] = HttpClientTransport.normalizeEvenLengthHex(signature, "opening.signature")
        return opening
    }
}
